lib/charms/operator_libs_linux/v0/sysctl.py

```diff
@@ -27,6 +87,211 @@
+
+        snapshot = self._create_snapshot()
+        logger.debug("Created snapshot for keys: %s", snapshot)
+        try:
+            self._apply()
+        except SysctlPermissionError:
+            self._restore_snapshot(snapshot)
+            raise
+
+        self._create_charm_file()
+        self._merge()
+
+    def remove(self) -> None:
+        """Remove config for charm."""
+        self.charm_filepath.unlink(missing_ok=True)
+        logger.info("Charm config file %s was removed", self.charm_filepath)
+        self._merge()
+
+    def _validate(self) -> List[str]:
+        """Validate the desired config params against merged ones."""
+        common_keys = set(self._data.keys()) & set(self._desired_config.keys())
+        conflict_keys = []
+        for key in common_keys:
+            if self._data[key] != self._desired_config[key]:
+                logger.warning(
+                    "Values for key '%s' are different: %s != %s",
+                    key,
+                    self._data[key],
+                    self._desired_config[key],
+                )
+                conflict_keys.append(key)
+
+        return conflict_keys
+
+    def _create_charm_file(self) -> None:
+        """Write the charm file."""
+        charm_params = [f"{key}={value}\n" for key, value in self._desired_config.items()]
+        with open(self.charm_filepath, "w") as f:
+            f.writelines(charm_params)
+
+    def _merge(self, add_own_charm=True) -> None:
+        """Create the merged sysctl file.
+
+        Args:
+            add_own_charm : bool, if false it will skip the charm file from the merge.
+        """
+        # get all files that start by 90-juju-
+        data = [SYSCTL_HEADER]
+        paths = set(SYSCTL_DIRECTORY.glob(f"{CHARM_FILENAME_PREFIX}*"))
+        if not add_own_charm:
+            paths.discard(self.charm_filepath.as_posix())
+
+        for path in paths:
+            with open(path, "r") as f:
+                data += f.readlines()
+        with open(SYSCTL_FILENAME, "w") as f:
+            f.writelines(data)
+
+        # Reload data with newly created file.
+        self._data = self._load_data()
+
+    def _apply(self) -> None:
+        """Apply values to machine."""
+        cmd = [f"{key}={value}" for key, value in self._desired_config.items()]
+        result = self._sysctl(cmd)
+        expr = re.compile(r"^sysctl: permission denied on key \"([a-z_\.]+)\", ignoring$")
+        failed_values = [expr.match(line) for line in result if expr.match(line)]
+        logger.debug("Failed values: %s", failed_values)
+
+        if failed_values:
+            msg = f"Unable to set params: {[f.group(1) for f in failed_values]}"
+            logger.error(msg)
+            raise SysctlPermissionError(msg)
+
+    def _create_snapshot(self) -> Dict[str, str]:
+        """Create a snaphot of config options that are going to be set."""
+        return {key: self._sysctl([key, "-n"])[0] for key in self._desired_config.keys()}
```

Comment:
Like this we will call subprocess to many times and this can easily took 10+ seconds. Can we call it only once or load those values from `95-juju-sysctl.cong`?